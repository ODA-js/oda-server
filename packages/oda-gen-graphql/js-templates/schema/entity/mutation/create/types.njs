<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-create-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./mutations/create/create${entity.name}Input.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input create#{entity.name}Input {
<#- for (let field of entity.create){#>
      #{field.name}: #{field.type}
<#-}#>
<#- for (let rel of entity.relations.filter(f=>f.persistent)){
const refName = rel.fields.length > 0 ? `embed${rel.ref.entity}CreateInto${entity.name}${rel.cField}Input` : `embed${rel.ref.entity}Input`;
#>
      #{rel.field}: <#if(!rel.single){#>[<#}#>#{refName}<#if(!rel.single){#>]<#}#>
<#-}#>
    }
  `,
});

<#- chunkStart(`./mutations/create/create${entity.name}Payload.ts`); -#>

import { Type } from '../../../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type create#{entity.name}Payload {
      #{entity.payloadName}: #{entity.plural}Edge
    }
  `,
});
