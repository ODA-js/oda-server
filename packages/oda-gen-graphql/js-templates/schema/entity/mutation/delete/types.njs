<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-delete-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./mutations/delete/delete${entity.name}Input.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input delete#{entity.name}Input {
<#- for (let field of entity.unique){#>
      #{field.name}: #{field.type}
<#-}#>
    } 
  `,
});

<#- chunkStart(`./mutations/delete/delete${entity.name}Payload.ts`); -#>

import { Type } from '../../../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type delete#{entity.name}Payload {
      deletedItemId: ID
      #{entity.payloadName}: #{entity.name}
    }
  `,
});
