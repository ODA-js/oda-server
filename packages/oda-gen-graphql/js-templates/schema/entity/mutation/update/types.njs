<#@ chunks "$$$main$$$" -#>
<#@ alias 'mutations-update-types'#>
<#@ context 'entity'#>

<#- chunkStart(`./mutations/update/update${entity.name}Input.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input update#{entity.name}Input {
    <#- for (let field of entity.update){#>
      #{field.name}: #{field.type}
    <#-}#>
    <#- for (let rel of entity.relations.filter(f=>f.persistent)){

    const createName = rel.fields.length > 0 ? `embed${rel.ref.entity}CreateInto${entity.name}${rel.cField}Input` : `create${rel.ref.entity}Input`;
    const updateName = rel.fields.length > 0 ? `embed${rel.ref.entity}UpdateInto${entity.name}${rel.cField}Input` : `embed${rel.ref.entity}Input`;
    #>
      #{rel.field}: <#if(!rel.single){#>[<#}#>#{updateName}<#if(!rel.single){#>]<#}#>
<#- if(!rel.embedded){#>
      #{rel.field}Unlink: <#if(!rel.single){#>[<#}#>#{updateName}<#if(!rel.single){#>]<#}#>
      #{rel.field}Create: <#if(!rel.single){#>[<#}#>#{createName}<#if(!rel.single){#>]<#}#>
<#-}#>
    <#-}#>
    }
  `,
});

<#- chunkEnd(); -#>

<#- for (let rel of entity.relations.filter(f=>f.persistent && f.fields.length > 0)){
const createName = `embed${rel.ref.entity}CreateInto${entity.name}${rel.cField}Input`;
const updateName = `embed${rel.ref.entity}UpdateInto${entity.name}${rel.cField}Input`;
let eEntity = rel.ref.eEntity;
  let eRels = eEntity.relations;
#>

<#- chunkStart(`./mutations/create/${createName}.ts`); -#>
<# slot('import-embed-entity-create-rel-mutation-types', createName) #>
<# slot('use-embed-entity-create-rel-mutation-types', createName) #>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
input #{createName} {
<#- for (let field of eEntity.create){#>
  #{field.name}: #{field.type}
<#-}#>
<#- for (let field of rel.fields){#>
  #{field.name}: #{field.type}
<#-}#>
<#-
 for (let eRel of eRels.filter(f=>f.persistent)){
const refName = eRel.fields.length > 0 ? `embed${eRel.ref.entity}UpdateInto${eEntity}${eRel.cField}Input` : `embed${eRel.ref.entity}Input`;
#>
  #{eRel.field}: <#if(!eRel.single){#>[<#}#>#{refName}<#if(!eRel.single){#>]<#}#>
<#-}#>
}`
});

<#- chunkEnd(); -#>
<#- chunkStart(`./mutations/update/${updateName}.ts`); -#>
<# slot('import-embed-entity-update-rel-mutation-types', updateName) #>
<# slot('use-embed-entity-update-rel-mutation-types', updateName) #>

import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
input #{updateName} {
<#- for (let field of eEntity.update){#>
  #{field.name}: #{field.type}
<#-}#>
<#- for (let field of rel.fields){#>
  #{field.name}: #{field.type}
<#-}#>
<#-
 for (let eRel of eRels.filter(f=>f.persistent)){
const refName = eRel.fields.length > 0 ? `embed${eRel.ref.entity}UpdateInto${eEntity}${eRel.cField}Input` : `embed${eRel.ref.entity}Input`;
#>
  #{eRel.field}: <#if(!eRel.single){#>[<#}#>#{refName}<#if(!eRel.single){#>]<#}#>
<#-}#>
}
`});
<#- chunkEnd(); -#>

<#-}#>


<#- chunkStart(`./mutations/update/update${entity.name}Payload.ts`); -#>

import { Type } from '../../../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type update#{entity.name}Payload {
      #{entity.payloadName}: #{entity.name}
    }
  `,
});
