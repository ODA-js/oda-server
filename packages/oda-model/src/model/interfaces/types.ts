export type RelationType = 'HasMany' | 'HasOne' | 'BelongsToMany' | 'BelongsTo';

export type MetaModelType =
  | 'metadata'
  | 'model'
  | 'query'
  | 'mutation'
  | 'package'
  | 'entity'
  | 'scalar'
  | 'mixin'
  | 'union'
  | 'enum'
  | 'enum-item'
  | 'field-base'
  | 'field'
  | 'enum-field'
  | 'entity-field'
  | 'relation-field'
  | 'relation'
  | 'operation'
  | 'ref'
  | RelationType;

export type Multiplicity = 'one' | 'many';
/**
 * kind of complex type
 */
export type ComplexTypeKind = 'enum' | 'entity';

export type OperationKind = 'create' | 'read' | 'update' | 'delete';

export type DirectiveLocation =
  | 'QUERY'
  | 'MUTATION'
  | 'SUBSCRIPTION'
  | 'FIELD'
  | 'FRAGMENT_DEFINITION'
  | 'FRAGMENT_SPREAD'
  | 'INLINE_FRAGMENT'
  | 'SCHEMA'
  | 'SCALAR'
  | 'OBJECT'
  | 'FIELD_DEFINITION'
  | 'ARGUMENT_DEFINITION'
  | 'INTERFACE'
  | 'UNION'
  | 'ENUM'
  | 'ENUM_VALUE'
  | 'INPUT_OBJECT'
  | 'INPUT_FIELD_DEFINITION';

export interface INamed {
  readonly name: string;
}

export interface FieldArgsForHash {
  type?: string;
  required?: boolean;
  defaultValue?: string;
}

export interface FieldArgs extends INamed, FieldArgsForHash {}
/**
 * Complext type can refer to existing type system
 */
export type EntityType = {
  /**
   * complext type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'entity';
  /**
   * multiplicity of it
   */
  multiplicity: Multiplicity;
};

export type EnumType = {
  /**
   * complext type name
   */
  name: string;
  /**
   * kind of complex type
   */
  type: 'enum';
  /**
   * multiplicity of it
   */
  multiplicity: Multiplicity;
};

export type FieldType = string | EnumType | EntityType;

export interface AsHash<T extends INamed> {
  [name: string]: T;
}

export function HashToMap<T extends INamed>(input: AsHash<T>): Map<string, T> {
  const res = Object.keys(input).map(name => {
    return [name, input[name] as T] as [string, T];
  });
  return new Map(res);
}

export function MapToHash<T extends INamed>(input: Map<string, T>): AsHash<T> {
  return [...input.entries()].reduce(
    (hash, [name, value]) => {
      hash[name] = value;
      return hash;
    },
    {} as AsHash<T>,
  );
}
