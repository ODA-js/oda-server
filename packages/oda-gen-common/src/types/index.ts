import { StateOfFieldType } from './stateOfField';
import { StateOfConectionType } from './stateOfConnection';
import { PageInfoType } from './pageInfo';
import { DateType } from './date';
import { IdType } from './id';
import { ImageSizeType } from './imageSize';
import { JSONType } from './json';
import { GQLModule } from './empty';
import { MutationKindType } from './mutationType';
import { WhereBoolean } from './whereBoolean';
import { WhereDate } from './whereDate';
import { WhereFloat } from './whereFloat';
import { WhereID } from './whereID';
import { WhereInt } from './whereInt';
import { WhereString } from './whereString';
import { WhereListOfStrings } from './whereListOfStrings';
import { WhereMutationKind } from './whereMutationKind';
import { WhereJSON } from './whereJSON';

export {
  GQLModule,
  StateOfFieldType,
  StateOfConectionType,
  PageInfoType,
  DateType,
  IdType,
  ImageSizeType,
  JSONType,
  WhereBoolean,
  WhereDate,
  WhereFloat,
  WhereID,
  WhereInt,
  WhereString,
  WhereListOfStrings,
  WhereMutationKind,
  WhereJSON,
};

export class DefaultTypes extends GQLModule {
  protected _name = 'DefaultTypes';
  protected _composite = [
    new StateOfFieldType({}),
    new StateOfConectionType({}),
    new PageInfoType({}),
    new DateType({}),
    new IdType({}),
    new ImageSizeType({}),
    new JSONType({}),
    new MutationKindType({}),
    new WhereBoolean({}),
    new WhereDate({}),
    new WhereFloat({}),
    new WhereID({}),
    new WhereInt({}),
    new WhereString({}),
    new WhereListOfStrings({}),
    new WhereMutationKind({}),
    new WhereJSON({}),
  ];
}
