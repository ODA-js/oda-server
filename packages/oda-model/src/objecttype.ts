import { ElementMetaInfo, Internal, MetaData } from './element';
import {
  IModelBase,
  ModelBaseInput,
  ModelBaseInternal,
  ModelBase,
  ModelBaseOutput,
} from './modelbase';
import {
  AsHash,
  MetaModelType,
  Nullable,
  assignValue,
  HashToArray,
  NamedArray,
  MapToArray,
  ArrayToMap,
  ArgumentKind,
} from './types';

import { merge } from 'lodash';
import * as inflected from 'inflected';
import {
  ObjectTypeFieldInput,
  ObjectTypeField,
  IObjectTypeField,
} from './objecttypefield';
import { UIView } from './entitybase';

export interface IObjectType
  extends IModelBase<ObjectTypeMetaInfo, ObjectTypeInput, ObjectTypeOutput> {
  readonly kind: ArgumentKind;
  readonly plural: string;
  readonly global: boolean;
  readonly titlePlural: string;
  readonly fields: Map<string, IObjectTypeField>;
}

export interface ObjectTypeMetaInfo extends ElementMetaInfo {
  titlePlural: string;
  global: boolean;
  name: {
    plural: string;
    singular: string;
  };
  UI: UIView;
}

export interface ObjectTypeInternal extends ModelBaseInternal {
  fields: Map<string, IObjectTypeField>;
  kind: ArgumentKind;
}

export interface ObjectTypeInput extends ModelBaseInput<ObjectTypeMetaInfo> {
  kind?: ArgumentKind;
  plural?: string;
  global?: boolean;
  titlePlural?: string;
  fields: AsHash<ObjectTypeFieldInput> | NamedArray<ObjectTypeFieldInput>;
}

export function isObjectTypeInput(inp: any): inp is ObjectTypeInput {
  return typeof inp === 'object' && inp.hasOwnProperty('fields');
}

export function isObjectType(inp: any): inp is IObjectType | ObjectType {
  return (
    typeof inp === 'object' &&
    (inp.modelType as MetaModelType) === 'object-type'
  );
}

export interface ObjectTypeOutput extends ModelBaseOutput<ObjectTypeMetaInfo> {
  fields: NamedArray<ObjectTypeFieldInput>;
  kind: ArgumentKind;
}

export const recordDefaultMetaInfo = {
  name: {},
  UI: {
    listName: [],
    quickSearch: [],
    hidden: [],
    edit: [],
    show: [],
    list: [],
    embedded: [],
  },
};
export const recordDefaultInput = {
  metadata: recordDefaultMetaInfo,
  exact: false,
};

export class ObjectType
  extends ModelBase<
    ObjectTypeMetaInfo,
    ObjectTypeInput,
    ObjectTypeInternal,
    ObjectTypeOutput
  >
  implements IObjectType {
  public get modelType(): MetaModelType {
    return 'object-type';
  }

  constructor(init: ObjectTypeInput) {
    super(merge({}, recordDefaultInput, init));
  }

  get kind(): ArgumentKind {
    return this[Internal].kind;
  }

  get plural(): string {
    return this.metadata.name.plural;
  }

  get titlePlural(): string {
    return this.metadata.titlePlural;
  }

  get global(): boolean {
    return this.metadata.global;
  }

  get fields(): Map<string, IObjectTypeField> {
    return this[Internal].fields;
  }

  public updateWith(input: Nullable<ObjectTypeInput>) {
    super.updateWith(input);

    assignValue<ObjectTypeInternal, ObjectTypeInput, ObjectTypeInput['name']>({
      src: this[Internal],
      input,
      field: 'name',
      effect: (src, value) => {
        src.name = value.trim();
        this.metadata.name.singular = src.name;
      },
      required: true,
    });

    assignValue<
      ObjectTypeMetaInfo,
      ObjectTypeInput,
      NonNullable<ObjectTypeInput['global']>
    >({
      src: this[MetaData],
      input,
      field: 'global',
      effect: (src, value) => (src.global = value),
      setDefault: src => (src.global = false),
    });

    assignValue<
      ObjectTypeMetaInfo,
      ObjectTypeInput,
      NonNullable<ObjectTypeInput['plural']>
    >({
      src: this.metadata,
      input,
      field: 'name.plural',
      inputField: 'plural',
      effect: (src, value) => {
        src.name.plural = value.trim();
        if (src.name.plural === this.name) {
          src.name.plural = `All${this.name}`;
        }
      },
      setDefault: src => {
        src.name.plural = inflected.pluralize(this.name);
        if (src.name.plural === this.name) {
          src.name.plural = `All${this.name}`;
        }
      },
    });

    assignValue<
      ObjectTypeMetaInfo,
      ObjectTypeInput,
      NonNullable<ObjectTypeInput['titlePlural']>
    >({
      src: this.metadata,
      input,
      field: 'titlePlural',
      effect: (src, value) => (src.titlePlural = value.trim()),
      setDefault: src => (src.titlePlural = this.plural),
    });

    assignValue<
      ObjectTypeInternal,
      ObjectTypeInput,
      NonNullable<ObjectTypeInput['kind']>
    >({
      src: this[Internal],
      input,
      field: 'kind',
      effect: (src, value) => (src.kind = value),
      setDefault: src => (src.kind = 'input'),
    });

    assignValue<
      ObjectTypeInternal,
      ObjectTypeInput,
      NonNullable<ObjectTypeInput['fields']>
    >({
      src: this[Internal],
      input,
      field: 'fields',
      effect: (src, value) => {
        const fields = ArrayToMap(
          Array.isArray(value) ? value : HashToArray(value),
          (fld, order) => {
            let field = new ObjectTypeField(
              merge({}, fld, {
                order,
                entity: this.name,
                kind: this.kind,
              }),
            );
            return field;
          },
          (obj: any, src: any) => obj.mergeWith(src.toObject()),
        );

        src.fields = fields;
      },
      setDefault: src => (src.fields = new Map()),
    });
  }

  public toObject(): ObjectTypeOutput {
    return merge({}, super.toObject(), {
      fields: MapToArray(this.fields, (name, value) => ({
        ...value.toObject(),
        name,
      })),
      kind: this[Internal].kind,
    } as Partial<ObjectTypeOutput>);
  }

  public mergeWith(payload: Nullable<ObjectTypeInput>) {
    super.mergeWith(payload);
  }
}
