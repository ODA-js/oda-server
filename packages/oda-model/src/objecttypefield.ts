import {
  ModelBase,
  IModelBase,
  ModelBaseInternal,
  ModelBaseInput,
  ModelBaseOutput,
  ModelBaseMetaInfo,
} from './modelbase';
import { merge, get } from 'lodash';
import {
  EnumType,
  Nullable,
  assignValue,
  MetaModelType,
  Multiplicity,
  EntityType,
  ArgumentKind,
} from './types';
import { Internal } from './element';
import {
  IObjectType,
  ObjectTypeInput,
  ObjectTypeOutput,
  isObjectType,
  isObjectTypeInput,
  ObjectType,
} from './objecttype';

export interface IObjectTypeField
  extends IModelBase<
    ObjectTypeFieldMeta,
    ObjectTypeFieldInput,
    ObjectTypeFieldOutput
  > {
  readonly type: string | EnumType | EntityType | IObjectType;
  readonly multiplicity?: Multiplicity;
  readonly defaultValue?: string;
  readonly order?: number;
  readonly required: boolean;
  readonly kind: ArgumentKind;
}

export interface ObjectTypeFieldMeta extends ModelBaseMetaInfo {
  defaultValue?: string;
  order: number;
  required: boolean;
  multiplicity?: Multiplicity;
}

export interface ObjectTypeFieldInternal extends ModelBaseInternal {
  type: string | EnumType | EntityType | IObjectType;
  multiplicity: Multiplicity;
  kind: ArgumentKind;
}

export interface ObjectTypeFieldInput
  extends ModelBaseInput<ObjectTypeFieldMeta> {
  type?: string | EnumType | EntityType | ObjectTypeInput;
  required?: boolean;
  defaultValue?: string;
  multiplicity?: Multiplicity;
  order?: number;
  kind?: ArgumentKind;
}

export interface ObjectTypeFieldOutput
  extends ModelBaseOutput<ObjectTypeFieldMeta> {
  type?: string | EnumType | EntityType | ObjectTypeOutput;
  multiplicity?: Multiplicity;
  order?: number;
  kind: ArgumentKind;
}

export const typeFieldDefaultMetaInfo = {};
export const TypeFieldDefaultInput = { metadata: typeFieldDefaultMetaInfo };

export class ObjectTypeField
  extends ModelBase<
    ObjectTypeFieldMeta,
    ObjectTypeFieldInput,
    ObjectTypeFieldInternal,
    ObjectTypeFieldOutput
  >
  implements IObjectTypeField {
  public get modelType(): MetaModelType {
    const internal = this[Internal];
    return typeof internal.type === 'string'
      ? 'argument-simple-field'
      : isObjectType(internal.type)
      ? 'argument-object-type'
      : internal.type.type === 'entity'
      ? 'argument-entity-field'
      : 'argument-enum-field';
  }

  get type(): string | EnumType | EntityType | IObjectType {
    return this[Internal].type;
  }

  // if it is the field is List of Items i.e. String[]
  get multiplicity(): Multiplicity {
    return get(this[Internal], 'multiplicity', 'one');
  }

  get defaultValue(): string | undefined {
    return get(this.metadata, 'defaultValue');
  }

  get kind(): ArgumentKind {
    return this[Internal].kind;
  }

  get order(): number {
    return this.metadata.order;
  }

  get required(): boolean {
    return this.metadata.required;
  }

  constructor(init: ObjectTypeFieldInput) {
    super(merge({}, TypeFieldDefaultInput, init));
  }

  public updateWith(input: Nullable<ObjectTypeFieldInput>) {
    super.updateWith(input);

    assignValue<
      ObjectTypeFieldInternal,
      ObjectTypeFieldInput,
      ObjectTypeFieldInput['name']
    >({
      src: this[Internal],
      input,
      field: 'name',
      effect: (src, value) => (src.name = value.trim()),
      required: true,
    });

    assignValue<
      ObjectTypeFieldInternal,
      ObjectTypeFieldInput,
      NonNullable<ObjectTypeFieldInput['type']>
    >({
      src: this[Internal],
      input,
      field: 'type',
      effect: (src, value) => {
        if (typeof value !== 'string' && !isObjectTypeInput(value)) {
          if (!value.multiplicity) {
            value.multiplicity = 'one';
          } else {
            src.multiplicity = value.multiplicity;
          }
        }
        if (isObjectTypeInput(value)) {
          src.type = new ObjectType(value);
        } else {
          src.type = value;
        }
      },
      setDefault: src => (src.type = 'string'),
    });

    assignValue<
      ObjectTypeFieldInternal,
      ObjectTypeFieldInput,
      NonNullable<ObjectTypeFieldInput['multiplicity']>
    >({
      src: this[Internal],
      input,
      field: 'multiplicity',
      required: true,
      effect: (src, value) => {
        if (this.modelType === 'argument-enum-field') {
          (src.type as EnumType).multiplicity = value ? 'many' : 'one';
        }
        if (this.modelType === 'argument-entity-field') {
          (src.type as EntityType).multiplicity = value ? 'many' : 'one';
        }
        src.multiplicity = value;
      },
      setDefault: src => {
        if (this.modelType === 'argument-enum-field') {
          src.multiplicity = (src.type as EnumType).multiplicity || 'one';
        }
        if (this.modelType === 'argument-entity-field') {
          src.multiplicity = (src.type as EntityType).multiplicity || 'one';
        } else {
          src.multiplicity = 'one';
        }
      },
    });

    assignValue<
      ObjectTypeFieldMeta,
      ObjectTypeFieldInput,
      NonNullable<ObjectTypeFieldInput['defaultValue']>
    >({
      src: this.metadata,
      input,
      field: 'defaultValue',
      effect: (src, value) => {
        src.defaultValue = value;
      },
    });

    assignValue<
      ObjectTypeFieldMeta,
      ObjectTypeFieldInput,
      NonNullable<ObjectTypeFieldInput['order']>
    >({
      src: this.metadata,
      input,
      field: 'order',
      effect: (src, value) => (src.order = value),
    });

    assignValue<
      ObjectTypeFieldInternal,
      ObjectTypeFieldInput,
      NonNullable<ObjectTypeFieldInput['kind']>
    >({
      src: this[Internal],
      input,
      field: 'kind',
      effect: (src, value) => (src.kind = value),
      setDefault: src => (src.kind = 'input'),
    });

    assignValue<
      ObjectTypeFieldMeta,
      ObjectTypeFieldInput,
      NonNullable<ObjectTypeFieldInput['required']>
    >({
      src: this.metadata,
      input,
      field: 'required',
      effect: (src, value) => (src.required = value),
      required: true,
      setDefault: src => (src.required = false),
    });
  }

  public toObject(): ObjectTypeFieldOutput {
    return merge({}, super.toObject(), {
      type: isObjectType(this.type) ? this.type.toObject() : this.type,
      multiplicity: this[Internal].multiplicity,
    } as Partial<ObjectTypeFieldOutput>);
  }

  public mergeWith(payload: Nullable<ObjectTypeFieldInput>) {
    super.mergeWith(payload);
  }
}
