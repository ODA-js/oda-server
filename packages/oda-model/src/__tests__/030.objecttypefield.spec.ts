import 'jest';
import { ObjectTypeField } from '../objecttypefield';
import { EnumType } from '../types';
import { ObjectTypeInput } from '../objecttype';

describe('ObjectTypeField', () => {
  it('default', () => {
    const res = new ObjectTypeField({
      name: 'A',
    });
    expect(res.name).toBe('A');
    expect(res.type).toMatchObject({ name: 'String', type: 'scalar' });
    expect(res.modelType).toBe('argument-simple-field');
    expect(res.multiplicity).toBe('one');
    expect(res).toMatchSnapshot('default');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });

  it('default value only for persistent fields', () => {
    const res = new ObjectTypeField({
      name: 'A',
      defaultValue: 'NAME_OF_DEFAULT_VALUE',
    });
    expect(res.name).toBe('A');
    expect(res.defaultValue).toBe('NAME_OF_DEFAULT_VALUE');
  });

  it('required by default is false', () => {
    const res = new ObjectTypeField({
      name: 'A',
    });
    expect(res.name).toBe('A');
    expect(res.required).toBeFalsy();
  });
  it('required is setup', () => {
    const res = new ObjectTypeField({
      name: 'A',
      required: true,
    });
    expect(res.name).toBe('A');
    expect(res.required).toBeTruthy();
    expect(res.toObject()).toMatchSnapshot('required');
  });

  it('default enum', () => {
    const res = new ObjectTypeField({
      name: 'A',
      type: {
        type: 'enum',
        multiplicity: 'many',
        name: 'ENUM1',
      },
    });
    expect(res.name).toBe('A');
    expect(res.modelType).toBe('argument-enum-field');
    expect(res.multiplicity).toBe('many');
    expect((res.type as EnumType).type).toBe('enum');
    expect(res).toMatchSnapshot('default');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });

  it('default enum list', () => {
    const res = new ObjectTypeField({
      name: 'A',
      multiplicity: 'many',
      type: {
        type: 'enum',
        name: 'ENUM1',
      },
    });
    expect(res.name).toBe('A');
    expect(res.modelType).toBe('argument-enum-field');
    expect(res.multiplicity).toBe('many');
    expect((res.type as EnumType).type).toBe('enum');
    expect((res.type as EnumType).multiplicity).toBe('many');
  });
  it('default enum multiplicity == one', () => {
    const res = new ObjectTypeField({
      name: 'A',
      type: {
        type: 'enum',
        name: 'ENUM1',
      },
    });
    expect(res.name).toBe('A');
    expect(res.modelType).toBe('argument-enum-field');
    expect(res.multiplicity).toBe('one');
    expect((res.type as EnumType).type).toBe('enum');
    expect((res.type as EnumType).multiplicity).toBe('one');
    expect(res).toMatchSnapshot('default');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
  it('ObjectType as field type ObjectType', () => {
    const res = new ObjectTypeField({
      name: 'A',
      type: {
        name: 'one',
        fields: [
          {
            name: 'a',
          },
          {
            name: 'b',
          },
          {
            name: 'c',
            type: {
              name: 'C',
              fields: [
                {
                  name: 'c1',
                },
              ],
            },
          },
        ],
      } as ObjectTypeInput,
    });
    expect(res).toMatchSnapshot('object');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
});
