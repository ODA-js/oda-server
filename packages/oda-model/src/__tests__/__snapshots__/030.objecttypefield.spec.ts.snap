// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectTypeField ObjectType as field type ObjectType: object 1`] = `
ObjectTypeField {
  Symbol(internal): Object {
    "kind": "input",
    "multiplicity": "one",
    "name": "A",
    "type": ObjectType {
      Symbol(internal): Object {
        "fields": Map {
          "a" => ObjectTypeField {
            Symbol(internal): Object {
              "kind": "input",
              "multiplicity": "one",
              "name": "a",
              "type": "String",
            },
            Symbol(metadata): Object {
              "order": 0,
              "required": false,
            },
          },
          "b" => ObjectTypeField {
            Symbol(internal): Object {
              "kind": "input",
              "multiplicity": "one",
              "name": "b",
              "type": "String",
            },
            Symbol(metadata): Object {
              "order": 1,
              "required": false,
            },
          },
          "c" => ObjectTypeField {
            Symbol(internal): Object {
              "kind": "input",
              "multiplicity": "one",
              "name": "c",
              "type": ObjectType {
                Symbol(internal): Object {
                  "fields": Map {
                    "c1" => ObjectTypeField {
                      Symbol(internal): Object {
                        "kind": "input",
                        "multiplicity": "one",
                        "name": "c1",
                        "type": "String",
                      },
                      Symbol(metadata): Object {
                        "order": 0,
                        "required": false,
                      },
                    },
                  },
                  "kind": "input",
                  "name": "C",
                },
                Symbol(metadata): Object {
                  "UI": Object {
                    "edit": Array [],
                    "embedded": Array [],
                    "hidden": Array [],
                    "list": Array [],
                    "listName": Array [],
                    "quickSearch": Array [],
                    "show": Array [],
                  },
                  "global": false,
                  "name": Object {
                    "plural": "Cs",
                    "singular": "C",
                  },
                  "titlePlural": "Cs",
                },
              },
            },
            Symbol(metadata): Object {
              "order": 2,
              "required": false,
            },
          },
        },
        "kind": "input",
        "name": "one",
      },
      Symbol(metadata): Object {
        "UI": Object {
          "edit": Array [],
          "embedded": Array [],
          "hidden": Array [],
          "list": Array [],
          "listName": Array [],
          "quickSearch": Array [],
          "show": Array [],
        },
        "global": false,
        "name": Object {
          "plural": "ones",
          "singular": "one",
        },
        "titlePlural": "ones",
      },
    },
  },
  Symbol(metadata): Object {
    "required": false,
  },
}
`;

exports[`ObjectTypeField ObjectType as field type ObjectType: toObject 1`] = `
Object {
  "description": undefined,
  "metadata": Object {
    "required": false,
  },
  "multiplicity": "one",
  "name": "A",
  "title": undefined,
  "type": Object {
    "description": undefined,
    "fields": Array [
      Object {
        "description": undefined,
        "metadata": Object {
          "order": 0,
          "required": false,
        },
        "multiplicity": "one",
        "name": "a",
        "title": undefined,
        "type": "String",
      },
      Object {
        "description": undefined,
        "metadata": Object {
          "order": 1,
          "required": false,
        },
        "multiplicity": "one",
        "name": "b",
        "title": undefined,
        "type": "String",
      },
      Object {
        "description": undefined,
        "metadata": Object {
          "order": 2,
          "required": false,
        },
        "multiplicity": "one",
        "name": "c",
        "title": undefined,
        "type": Object {
          "description": undefined,
          "fields": Array [
            Object {
              "description": undefined,
              "metadata": Object {
                "order": 0,
                "required": false,
              },
              "multiplicity": "one",
              "name": "c1",
              "title": undefined,
              "type": "String",
            },
          ],
          "kind": "input",
          "metadata": Object {
            "UI": Object {
              "edit": Array [],
              "embedded": Array [],
              "hidden": Array [],
              "list": Array [],
              "listName": Array [],
              "quickSearch": Array [],
              "show": Array [],
            },
            "global": false,
            "name": Object {
              "plural": "Cs",
              "singular": "C",
            },
            "titlePlural": "Cs",
          },
          "name": "C",
          "title": undefined,
        },
      },
    ],
    "kind": "input",
    "metadata": Object {
      "UI": Object {
        "edit": Array [],
        "embedded": Array [],
        "hidden": Array [],
        "list": Array [],
        "listName": Array [],
        "quickSearch": Array [],
        "show": Array [],
      },
      "global": false,
      "name": Object {
        "plural": "ones",
        "singular": "one",
      },
      "titlePlural": "ones",
    },
    "name": "one",
    "title": undefined,
  },
}
`;

exports[`ObjectTypeField default enum multiplicity == one: default 1`] = `
ObjectTypeField {
  Symbol(internal): Object {
    "kind": "input",
    "multiplicity": "one",
    "name": "A",
    "type": Object {
      "multiplicity": "one",
      "name": "ENUM1",
      "type": "enum",
    },
  },
  Symbol(metadata): Object {
    "required": false,
  },
}
`;

exports[`ObjectTypeField default enum multiplicity == one: toObject 1`] = `
Object {
  "description": undefined,
  "metadata": Object {
    "required": false,
  },
  "multiplicity": "one",
  "name": "A",
  "title": undefined,
  "type": Object {
    "multiplicity": "one",
    "name": "ENUM1",
    "type": "enum",
  },
}
`;

exports[`ObjectTypeField default enum: default 1`] = `
ObjectTypeField {
  Symbol(internal): Object {
    "kind": "input",
    "multiplicity": "many",
    "name": "A",
    "type": Object {
      "multiplicity": "many",
      "name": "ENUM1",
      "type": "enum",
    },
  },
  Symbol(metadata): Object {
    "required": false,
  },
}
`;

exports[`ObjectTypeField default enum: toObject 1`] = `
Object {
  "description": undefined,
  "metadata": Object {
    "required": false,
  },
  "multiplicity": "many",
  "name": "A",
  "title": undefined,
  "type": Object {
    "multiplicity": "many",
    "name": "ENUM1",
    "type": "enum",
  },
}
`;

exports[`ObjectTypeField default: default 1`] = `
ObjectTypeField {
  Symbol(internal): Object {
    "kind": "input",
    "multiplicity": "one",
    "name": "A",
    "type": "String",
  },
  Symbol(metadata): Object {
    "required": false,
  },
}
`;

exports[`ObjectTypeField default: toObject 1`] = `
Object {
  "description": undefined,
  "metadata": Object {
    "required": false,
  },
  "multiplicity": "one",
  "name": "A",
  "title": undefined,
  "type": "String",
}
`;

exports[`ObjectTypeField required is setup: required 1`] = `
Object {
  "description": undefined,
  "metadata": Object {
    "required": true,
  },
  "multiplicity": "one",
  "name": "A",
  "title": undefined,
  "type": "String",
}
`;
